### PR TITLE
Added support for space variables

### PR DESCRIPTION
**Pull request by martinbrose**
Addresses:

- #1552

### CODE REVIEW

File: src/huggingface_hub/_space_api.py [hunk 1]
```diff
@@ -118,3 +121,31 @@ def __init__(self, data: Dict) -> None:
...
+        description (`str` or None):
+            Description of the variable. Example: `"Model Repo ID of the implemented model"`.
+        updatedAt (`datetime`):
+            datetime of the last update of the variable.
+    """
+
+    key: str
+    value: str
+    description: Optional[str]
+    updatedAt: datetime
```
Comment (Wauplin) [inline src/huggingface_hub/_space_api.py:145]:
```suggestion
    updated_at: datetime
```

For consistency, let's use `snake_case` for all attributes

File: src/huggingface_hub/_space_api.py [hunk 2]
```diff
@@ -118,3 +121,31 @@ def __init__(self, data: Dict) -> None:
...
+
+    def __init__(self, k, v: Dict) -> None:
+        self.key = k
+        self.value = v["value"]
+        self.description = v["description"]
+        self.updatedAt = parse_datetime(v["updatedAt"])
```
Comment (Wauplin) [inline src/huggingface_hub/_space_api.py:151]:
```suggestion
    def __init__(self, key: str, values: Dict) -> None:
        self.key = key
        self.value = values["value"]
        self.description = values.get("description")
        self.updated_at = parse_datetime(v["updatedAt"])
```

- `description` is optional. If it is not returned by the server, we should set it to None
- best to explicit the parameters (k => key, v => values) + best to explicit the key type (str)
- updatedAt => updated_at

File: src/huggingface_hub/hf_api.py [hunk 3]
```diff
@@ -5020,6 +5020,79 @@ def delete_space_secret(self, repo_id: str, key: str, *, token: Optional[str] =
         )
         hf_raise_for_status(r)
 
+    @validate_hf_hub_args
+    def get_space_variables(self, repo_id: str, *, token: Optional[str] = None) -> Iterable[SpaceVariable]:
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:5024]:
```suggestion
    def get_space_variables(self, repo_id: str, *, token: Optional[str] = None) -> List[SpaceVariable]:
```

Choosing between returning an Iterable (a generator) or a List can be tricky. The main differences I see are:
- Lists are most practical to use. Users can print them, loop through them multiple times, access a specific item,...
- Iterables can be more memory efficient since the list is not stored in memory but fetched only on request. But this is true only when the data is huge and you need multiple calls to the Hub to load them. For example, `list_models` returns an Iterable because we fetch the results page after page. In our case here, all the environment variables are fetched at once so we won't gain any performance.

In general the rule is: use List whenever possible except if the result can be lazy-loaded.

Here I would go for a List.

File: src/huggingface_hub/hf_api.py [hunk 4]
```diff
@@ -5020,6 +5020,79 @@ def delete_space_secret(self, repo_id: str, key: str, *, token: Optional[str] =
...
+            token (`str`, *optional*):
+                Hugging Face token. Will default to the locally saved token if not provided.
+        """
+        r = get_session().get(
+            f"{self.endpoint}/api/spaces/{repo_id}/variables",
+            headers=self._build_hf_headers(token=token),
+        )
+        hf_raise_for_status(r)
+        for k, v in r.json().items():
+            yield (SpaceVariable(k, v))
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:5042]:
```suggestion
        return [SpaceVariable(k, v) for k, v in r.json().items()]
```

Better to return a list instead of a generator (see comment above).

File: src/huggingface_hub/hf_api.py [hunk 5]
```diff
@@ -5020,6 +5020,79 @@ def delete_space_secret(self, repo_id: str, key: str, *, token: Optional[str] =
...
+
+    @validate_hf_hub_args
+    def add_space_variable(
+        self, repo_id: str, key: str, value: str, description: Optional[str] = str(), *, token: Optional[str] = None
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:5046]:
```suggestion
        self, repo_id: str, key: str, value: str, *, description: Optional[str] = None, token: Optional[str] = None
```

Same comment as https://github.com/huggingface/huggingface_hub/pull/1594#discussion_r1295938121. I'd prefer keywork-only parameter + it's best to default to None when the value is not known instead of `""`.

File: src/huggingface_hub/hf_api.py [hunk 6]
```diff
@@ -5020,6 +5020,79 @@ def delete_space_secret(self, repo_id: str, key: str, *, token: Optional[str] =
...
+            description (`str`):
+                Description of the variable. Example: `"Model Repo ID of the implemented model"`.
+            token (`str`, *optional*):
+                Hugging Face token. Will default to the locally saved token if not provided.
+        """
+        r = get_session().post(
+            f"{self.endpoint}/api/spaces/{repo_id}/variables",
+            headers=self._build_hf_headers(token=token),
+            json={"key": key, "value": value, "description": description},
+        )
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:5069]:
```suggestion
        payload = {"key": key, "value": value}
        if description is not None:
            payload["description"] = description
        r = get_session().post(
            f"{self.endpoint}/api/spaces/{repo_id}/variables", headers=self._build_hf_headers(token=token), json=payload
        )
```

Better to send the description to the server only when the user has explicitly set the value.

File: src/huggingface_hub/hf_api.py (hunk 4 above)
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py:5042]: Thinking twice about it, it would be best to return a dictionary. It will be easier to work with for users for example to do `if "MODEL_REPO_ID" in get_space_variables(...): ...`. This applies to all methods that return the space variables.

```suggestion
        return {k: SpaceVariable(k, v) for k, v in r.json().items()}
```

**Review by Wauplin — CHANGES_REQUESTED** [review]
Very clean PR @martinbrose! I've added some comment to fix a few corner cases but the logic+docstring+tests looks good overall. Happy to make another review once those are addressed.

File: src/huggingface_hub/_space_api.py (hunk 2 above)
Comment (martinbrose) [reply src/huggingface_hub/_space_api.py:151]: The `values.get("description")` produces an exception for me.

I'll change it to:
```
if "description" in values.keys():
            self.description = values["description"]
```

File: src/huggingface_hub/hf_api.py (hunk 3 above)
Comment (martinbrose) [reply src/huggingface_hub/hf_api.py:5024]: With the below changes to return the variables as a dictionary, I believe I had to change it to `Dict[str, SpaceVariable]`?
I had related error messages when running `make quality`.

Comment (Wauplin) [reply src/huggingface_hub/hf_api.py]: > With the below changes to return the variables as a dictionary, I believe I had to change it to Dict[str, SpaceVariable]?

Yes exactly!

File: src/huggingface_hub/_space_api.py (hunk 2 above)
Comment (Wauplin) [reply src/huggingface_hub/_space_api.py]: Hmmm, which error are you getting with `values.get("description")` ? Your solution is valid as well (you can even do `if "description" in values:` instead) but using `.get(...)` should be equivalent.

File: src/huggingface_hub/__init__.py [hunk 7]
```diff
@@ -142,7 +142,7 @@
         "RepoUrl",
         "UserLikes",
         "add_space_secret",
-        "change_discussion_status",
```
Comment (Wauplin) [inline src/huggingface_hub/__init__.py:145]: Typo on this line. You might have forgotten a comma `,` between two lines and the linting merged the two strings in one.

File: src/huggingface_hub/_space_api.py (hunk 2 above)
Comment (martinbrose) [reply src/huggingface_hub/_space_api.py]: A good question. I just tried to replicate it and somehow can't. 
I must have made an error previously.

Changed to your proposal of `values.get("description")` because it's less code.

File: src/huggingface_hub/__init__.py (hunk 7 above)
Comment (martinbrose) [reply src/huggingface_hub/__init__.py:145]: Oh, yes. Apologies!

The same error was also present a little bit further down. Corrected as well.

**Comment by martinbrose** [comment]
Hi @Wauplin, I believe I have addressed all review comments.

Please check once you have a chance.

**Review by Wauplin — COMMENTED** [review]
Everything looks good to me!

One last thing: would you mind adding `SpaceVariable` object to the dataclasses list in this [reference doc](https://huggingface.co/docs/huggingface_hub/v0.16.3/en/package_reference/space_runtime#data-structures)? You can do that by adding a section and `[[autodoc]] SpaceVariable` [in this file](https://github.com/huggingface/huggingface_hub/blob/main/docs/source/en/package_reference/space_runtime.md#data-structures). You will also have to add `SpaceVariable` to the [`__init__.py` file](https://github.com/huggingface/huggingface_hub/blob/da50c580f01a01fa1e4934e45d3b5efc89cc0af1/src/huggingface_hub/__init__.py#L71) and run `make style`.

**Comment by martinbrose** [comment]
@Wauplin Done.

**Review by Wauplin — APPROVED** [review]
